diff --git a/components/board.tsx b/components/board.tsx
index 57e2531..e24489f 100644
--- a/components/board.tsx
+++ b/components/board.tsx
@@ -12,9 +12,16 @@ export default function Board() {
 
   const handleKeyDown = useCallback(
     (e: KeyboardEvent) => {
-      // disables page scrolling with keyboard arrows
+      if (!e.code.startsWith("Arrow")) {
+        return;
+      }
+
       e.preventDefault();
 
+      if (status !== "ongoing") {
+        return;
+      }
+
       switch (e.code) {
         case "ArrowUp":
           moveTiles("move_up");
@@ -30,11 +37,15 @@ export default function Board() {
           break;
       }
     },
-    [moveTiles],
+    [moveTiles, status],
   );
 
   const handleSwipe = useCallback(
     ({ deltaX, deltaY }: SwipeInput) => {
+      if (status !== "ongoing") {
+        return;
+      }
+
       if (Math.abs(deltaX) > Math.abs(deltaY)) {
         if (deltaX > 0) {
           moveTiles("move_right");
@@ -49,7 +60,7 @@ export default function Board() {
         }
       }
     },
-    [moveTiles],
+    [moveTiles, status],
   );
 
   const renderGrid = () => {
@@ -87,8 +98,7 @@ export default function Board() {
   return (
     <MobileSwiper onSwipe={handleSwipe}>
       <div className={styles.board}>
-        {status === "won" && <Splash heading="You won!" type="won" />}
-        {status === "lost" && <Splash heading="You lost!" />}
+        <Splash />
         <div className={styles.tiles}>{renderTiles()}</div>
         <div className={styles.grid}>{renderGrid()}</div>
       </div>
diff --git a/components/splash.tsx b/components/splash.tsx
index de37256..2430498 100644
--- a/components/splash.tsx
+++ b/components/splash.tsx
@@ -1,18 +1,19 @@
 import { GameContext } from "@/context/game-context";
-import styles from "@/styles/splash.module.css";
 import { useContext } from "react";
+import GameStatusModal from "./GameStatusModal";
 
-export default function Splash({ heading = "You won!", type = "" }) {
-  const { startGame } = useContext(GameContext);
+export default function Splash() {
+  const { continueGame, startGame, status } = useContext(GameContext);
+
+  if (status === "ongoing") {
+    return null;
+  }
 
   return (
-    <div className={`${styles.splash} ${type === "won" && styles.win}`}>
-      <div>
-        <h1>{heading}</h1>
-        <button className={styles.button} onClick={startGame}>
-          Play again
-        </button>
-      </div>
-    </div>
+    <GameStatusModal
+      status={status}
+      onRestart={startGame}
+      onContinue={status === "won" ? continueGame : undefined}
+    />
   );
 }
diff --git a/context/game-context.tsx b/context/game-context.tsx
index 0b93edc..6b18a38 100644
--- a/context/game-context.tsx
+++ b/context/game-context.tsx
@@ -6,22 +6,28 @@ import {
   useReducer,
 } from "react";
 import { isNil, throttle } from "lodash";
-import {
-  gameWinTileValue,
-  mergeAnimationDuration,
-  tileCountPerDimension,
-} from "@/constants";
+import { mergeAnimationDuration, tileCountPerDimension } from "@/constants";
 import { Tile } from "@/models/tile";
-import gameReducer, { initialState } from "@/reducers/game-reducer";
+import gameReducer, { GameStatus, initialState } from "@/reducers/game-reducer";
 
 type MoveDirection = "move_up" | "move_down" | "move_left" | "move_right";
 
-export const GameContext = createContext({
+type GameContextValue = {
+  score: number;
+  status: GameStatus;
+  moveTiles: (direction: MoveDirection) => void;
+  getTiles: () => Tile[];
+  startGame: () => void;
+  continueGame: () => void;
+};
+
+export const GameContext = createContext<GameContextValue>({
   score: 0,
   status: "ongoing",
   moveTiles: (_: MoveDirection) => {},
   getTiles: () => [] as Tile[],
   startGame: () => {},
+  continueGame: () => {},
 });
 
 export default function GameProvider({ children }: PropsWithChildren) {
@@ -37,17 +43,20 @@ export default function GameProvider({ children }: PropsWithChildren) {
         }
       }
     }
+
     return results;
   };
 
   const appendRandomTile = () => {
     const emptyCells = getEmptyCells();
+
     if (emptyCells.length > 0) {
       const cellIndex = Math.floor(Math.random() * emptyCells.length);
       const newTile = {
         position: emptyCells[cellIndex],
         value: 2,
       };
+
       dispatch({ type: "create_tile", tile: newTile });
     }
   };
@@ -58,11 +67,17 @@ export default function GameProvider({ children }: PropsWithChildren) {
 
   const moveTiles = useCallback(
     throttle(
-      (type: MoveDirection) => dispatch({ type }),
+      (type: MoveDirection) => {
+        if (gameState.status !== "ongoing") {
+          return;
+        }
+
+        dispatch({ type });
+      },
       mergeAnimationDuration * 1.05,
       { trailing: false },
     ),
-    [dispatch],
+    [gameState.status],
   );
 
   const startGame = () => {
@@ -71,55 +86,23 @@ export default function GameProvider({ children }: PropsWithChildren) {
     dispatch({ type: "create_tile", tile: { position: [0, 2], value: 2 } });
   };
 
-  const checkGameState = () => {
-    const isWon =
-      Object.values(gameState.tiles).filter((t) => t.value === gameWinTileValue)
-        .length > 0;
-
-    if (isWon) {
-      dispatch({ type: "update_status", status: "won" });
-      return;
-    }
-
-    const { tiles, board } = gameState;
-
-    const maxIndex = tileCountPerDimension - 1;
-    for (let x = 0; x < maxIndex; x += 1) {
-      for (let y = 0; y < maxIndex; y += 1) {
-        if (
-          isNil(gameState.board[x][y]) ||
-          isNil(gameState.board[x + 1][y]) ||
-          isNil(gameState.board[x][y + 1])
-        ) {
-          return;
-        }
-
-        if (tiles[board[x][y]].value === tiles[board[x + 1][y]].value) {
-          return;
-        }
-
-        if (tiles[board[x][y]].value === tiles[board[x][y + 1]].value) {
-          return;
-        }
-      }
-    }
-
-    dispatch({ type: "update_status", status: "lost" });
+  const continueGame = () => {
+    dispatch({ type: "continue_game" });
   };
 
-  useEffect(() => {
-    if (gameState.hasChanged) {
-      setTimeout(() => {
-        dispatch({ type: "clean_up" });
-        appendRandomTile();
-      }, mergeAnimationDuration);
-    }
-  }, [gameState.hasChanged]);
-
   useEffect(() => {
     if (!gameState.hasChanged) {
-      checkGameState();
+      return;
     }
+
+    const timeoutId = window.setTimeout(() => {
+      dispatch({ type: "clean_up" });
+      appendRandomTile();
+    }, mergeAnimationDuration);
+
+    return () => {
+      window.clearTimeout(timeoutId);
+    };
   }, [gameState.hasChanged]);
 
   return (
@@ -130,6 +113,7 @@ export default function GameProvider({ children }: PropsWithChildren) {
         getTiles,
         moveTiles,
         startGame,
+        continueGame,
       }}
     >
       {children}
diff --git a/reducers/game-reducer.ts b/reducers/game-reducer.ts
index 47d6226..f1db0ed 100644
--- a/reducers/game-reducer.ts
+++ b/reducers/game-reducer.ts
@@ -1,18 +1,20 @@
 import { flattenDeep, isEqual, isNil } from "lodash";
 import { uid } from "uid";
-import { tileCountPerDimension } from "@/constants";
+import { gameWinTileValue, tileCountPerDimension } from "@/constants";
 import { Tile, TileMap } from "@/models/tile";
 
-type GameStatus = "ongoing" | "won" | "lost";
+export type GameStatus = "ongoing" | "won" | "lost";
 
-type State = {
+export type State = {
   board: string[][];
   tiles: TileMap;
   tilesByIds: string[];
   hasChanged: boolean;
   score: number;
   status: GameStatus;
+  hasWon: boolean;
 };
+
 type Action =
   | { type: "create_tile"; tile: Tile }
   | { type: "clean_up" }
@@ -21,6 +23,7 @@ type Action =
   | { type: "move_left" }
   | { type: "move_right" }
   | { type: "reset_game" }
+  | { type: "continue_game" }
   | { type: "update_status"; status: GameStatus };
 
 function createBoard() {
@@ -33,6 +36,283 @@ function createBoard() {
   return board;
 }
 
+function hasWinningTile(tiles: TileMap) {
+  return Object.values(tiles).some((tile) => tile.value >= gameWinTileValue);
+}
+
+export function isGameOver(board: string[][], tiles: TileMap) {
+  for (let y = 0; y < tileCountPerDimension; y += 1) {
+    for (let x = 0; x < tileCountPerDimension; x += 1) {
+      const tileId = board[y][x];
+
+      if (isNil(tileId)) {
+        return false;
+      }
+
+      const currentValue = tiles[tileId].value;
+
+      if (x < tileCountPerDimension - 1) {
+        const rightTileId = board[y][x + 1];
+
+        if (!isNil(rightTileId) && tiles[rightTileId].value === currentValue) {
+          return false;
+        }
+      }
+
+      if (y < tileCountPerDimension - 1) {
+        const bottomTileId = board[y + 1][x];
+
+        if (!isNil(bottomTileId) && tiles[bottomTileId].value === currentValue) {
+          return false;
+        }
+      }
+    }
+  }
+
+  return true;
+}
+
+function resolveStatus(state: State, nextState: State): State {
+  if (!nextState.hasWon && hasWinningTile(nextState.tiles)) {
+    return {
+      ...nextState,
+      status: "won",
+      hasWon: true,
+    };
+  }
+
+  if (state.status === "won") {
+    return {
+      ...nextState,
+      status: "won",
+    };
+  }
+
+  if (isGameOver(nextState.board, nextState.tiles)) {
+    return {
+      ...nextState,
+      status: "lost",
+    };
+  }
+
+  return {
+    ...nextState,
+    status: state.status === "lost" ? "lost" : "ongoing",
+  };
+}
+
+function moveUp(state: State) {
+  const newBoard = createBoard();
+  const newTiles: TileMap = {};
+  let hasChanged = false;
+  let { score } = state;
+
+  for (let x = 0; x < tileCountPerDimension; x++) {
+    let newY = 0;
+    let previousTile: Tile | undefined;
+
+    for (let y = 0; y < tileCountPerDimension; y++) {
+      const tileId = state.board[y][x];
+      const currentTile = state.tiles[tileId];
+
+      if (!isNil(tileId)) {
+        if (previousTile?.value === currentTile.value) {
+          score += previousTile.value * 2;
+          newTiles[previousTile.id as string] = {
+            ...previousTile,
+            value: previousTile.value * 2,
+          };
+          newTiles[tileId] = {
+            ...currentTile,
+            position: [x, newY - 1],
+          };
+          previousTile = undefined;
+          hasChanged = true;
+          continue;
+        }
+
+        newBoard[newY][x] = tileId;
+        newTiles[tileId] = {
+          ...currentTile,
+          position: [x, newY],
+        };
+        previousTile = newTiles[tileId];
+        if (!isEqual(currentTile.position, [x, newY])) {
+          hasChanged = true;
+        }
+        newY++;
+      }
+    }
+  }
+
+  return resolveStatus(state, {
+    ...state,
+    board: newBoard,
+    tiles: newTiles,
+    hasChanged,
+    score,
+  });
+}
+
+function moveDown(state: State) {
+  const newBoard = createBoard();
+  const newTiles: TileMap = {};
+  let hasChanged = false;
+  let { score } = state;
+
+  for (let x = 0; x < tileCountPerDimension; x++) {
+    let newY = tileCountPerDimension - 1;
+    let previousTile: Tile | undefined;
+
+    for (let y = tileCountPerDimension - 1; y >= 0; y--) {
+      const tileId = state.board[y][x];
+      const currentTile = state.tiles[tileId];
+
+      if (!isNil(tileId)) {
+        if (previousTile?.value === currentTile.value) {
+          score += previousTile.value * 2;
+          newTiles[previousTile.id as string] = {
+            ...previousTile,
+            value: previousTile.value * 2,
+          };
+          newTiles[tileId] = {
+            ...currentTile,
+            position: [x, newY + 1],
+          };
+          previousTile = undefined;
+          hasChanged = true;
+          continue;
+        }
+
+        newBoard[newY][x] = tileId;
+        newTiles[tileId] = {
+          ...currentTile,
+          position: [x, newY],
+        };
+        previousTile = newTiles[tileId];
+        if (!isEqual(currentTile.position, [x, newY])) {
+          hasChanged = true;
+        }
+        newY--;
+      }
+    }
+  }
+
+  return resolveStatus(state, {
+    ...state,
+    board: newBoard,
+    tiles: newTiles,
+    hasChanged,
+    score,
+  });
+}
+
+function moveLeft(state: State) {
+  const newBoard = createBoard();
+  const newTiles: TileMap = {};
+  let hasChanged = false;
+  let { score } = state;
+
+  for (let y = 0; y < tileCountPerDimension; y++) {
+    let newX = 0;
+    let previousTile: Tile | undefined;
+
+    for (let x = 0; x < tileCountPerDimension; x++) {
+      const tileId = state.board[y][x];
+      const currentTile = state.tiles[tileId];
+
+      if (!isNil(tileId)) {
+        if (previousTile?.value === currentTile.value) {
+          score += previousTile.value * 2;
+          newTiles[previousTile.id as string] = {
+            ...previousTile,
+            value: previousTile.value * 2,
+          };
+          newTiles[tileId] = {
+            ...currentTile,
+            position: [newX - 1, y],
+          };
+          previousTile = undefined;
+          hasChanged = true;
+          continue;
+        }
+
+        newBoard[y][newX] = tileId;
+        newTiles[tileId] = {
+          ...currentTile,
+          position: [newX, y],
+        };
+        previousTile = newTiles[tileId];
+        if (!isEqual(currentTile.position, [newX, y])) {
+          hasChanged = true;
+        }
+        newX++;
+      }
+    }
+  }
+
+  return resolveStatus(state, {
+    ...state,
+    board: newBoard,
+    tiles: newTiles,
+    hasChanged,
+    score,
+  });
+}
+
+function moveRight(state: State) {
+  const newBoard = createBoard();
+  const newTiles: TileMap = {};
+  let hasChanged = false;
+  let { score } = state;
+
+  for (let y = 0; y < tileCountPerDimension; y++) {
+    let newX = tileCountPerDimension - 1;
+    let previousTile: Tile | undefined;
+
+    for (let x = tileCountPerDimension - 1; x >= 0; x--) {
+      const tileId = state.board[y][x];
+      const currentTile = state.tiles[tileId];
+
+      if (!isNil(tileId)) {
+        if (previousTile?.value === currentTile.value) {
+          score += previousTile.value * 2;
+          newTiles[previousTile.id as string] = {
+            ...previousTile,
+            value: previousTile.value * 2,
+          };
+          newTiles[tileId] = {
+            ...currentTile,
+            position: [newX + 1, y],
+          };
+          previousTile = undefined;
+          hasChanged = true;
+          continue;
+        }
+
+        newBoard[y][newX] = tileId;
+        newTiles[tileId] = {
+          ...state.tiles[tileId],
+          position: [newX, y],
+        };
+        previousTile = newTiles[tileId];
+        if (!isEqual(currentTile.position, [newX, y])) {
+          hasChanged = true;
+        }
+        newX--;
+      }
+    }
+  }
+
+  return resolveStatus(state, {
+    ...state,
+    board: newBoard,
+    tiles: newTiles,
+    hasChanged,
+    score,
+  });
+}
+
 export const initialState: State = {
   board: createBoard(),
   tiles: {},
@@ -40,6 +320,7 @@ export const initialState: State = {
   hasChanged: false,
   score: 0,
   status: "ongoing",
+  hasWon: false,
 };
 
 export default function gameReducer(
@@ -76,7 +357,7 @@ export default function gameReducer(
       const newBoard = JSON.parse(JSON.stringify(state.board));
       newBoard[y][x] = tileId;
 
-      return {
+      return resolveStatus(state, {
         ...state,
         board: newBoard,
         tiles: {
@@ -87,214 +368,43 @@ export default function gameReducer(
           },
         },
         tilesByIds: [...state.tilesByIds, tileId],
-      };
+      });
     }
     case "move_up": {
-      const newBoard = createBoard();
-      const newTiles: TileMap = {};
-      let hasChanged = false;
-      let { score } = state;
-
-      for (let x = 0; x < tileCountPerDimension; x++) {
-        let newY = 0;
-        let previousTile: Tile | undefined;
-
-        for (let y = 0; y < tileCountPerDimension; y++) {
-          const tileId = state.board[y][x];
-          const currentTile = state.tiles[tileId];
-
-          if (!isNil(tileId)) {
-            if (previousTile?.value === currentTile.value) {
-              score += previousTile.value * 2;
-              newTiles[previousTile.id as string] = {
-                ...previousTile,
-                value: previousTile.value * 2,
-              };
-              newTiles[tileId] = {
-                ...currentTile,
-                position: [x, newY - 1],
-              };
-              previousTile = undefined;
-              hasChanged = true;
-              continue;
-            }
-
-            newBoard[newY][x] = tileId;
-            newTiles[tileId] = {
-              ...currentTile,
-              position: [x, newY],
-            };
-            previousTile = newTiles[tileId];
-            if (!isEqual(currentTile.position, [x, newY])) {
-              hasChanged = true;
-            }
-            newY++;
-          }
-        }
+      if (state.status !== "ongoing") {
+        return state;
       }
-      return {
-        ...state,
-        board: newBoard,
-        tiles: newTiles,
-        hasChanged,
-        score,
-      };
+
+      return moveUp(state);
     }
     case "move_down": {
-      const newBoard = createBoard();
-      const newTiles: TileMap = {};
-      let hasChanged = false;
-      let { score } = state;
-
-      for (let x = 0; x < tileCountPerDimension; x++) {
-        let newY = tileCountPerDimension - 1;
-        let previousTile: Tile | undefined;
-
-        for (let y = tileCountPerDimension - 1; y >= 0; y--) {
-          const tileId = state.board[y][x];
-          const currentTile = state.tiles[tileId];
-
-          if (!isNil(tileId)) {
-            if (previousTile?.value === currentTile.value) {
-              score += previousTile.value * 2;
-              newTiles[previousTile.id as string] = {
-                ...previousTile,
-                value: previousTile.value * 2,
-              };
-              newTiles[tileId] = {
-                ...currentTile,
-                position: [x, newY + 1],
-              };
-              previousTile = undefined;
-              hasChanged = true;
-              continue;
-            }
-
-            newBoard[newY][x] = tileId;
-            newTiles[tileId] = {
-              ...currentTile,
-              position: [x, newY],
-            };
-            previousTile = newTiles[tileId];
-            if (!isEqual(currentTile.position, [x, newY])) {
-              hasChanged = true;
-            }
-            newY--;
-          }
-        }
+      if (state.status !== "ongoing") {
+        return state;
       }
-      return {
-        ...state,
-        board: newBoard,
-        tiles: newTiles,
-        hasChanged,
-        score,
-      };
+
+      return moveDown(state);
     }
     case "move_left": {
-      const newBoard = createBoard();
-      const newTiles: TileMap = {};
-      let hasChanged = false;
-      let { score } = state;
-
-      for (let y = 0; y < tileCountPerDimension; y++) {
-        let newX = 0;
-        let previousTile: Tile | undefined;
-
-        for (let x = 0; x < tileCountPerDimension; x++) {
-          const tileId = state.board[y][x];
-          const currentTile = state.tiles[tileId];
-
-          if (!isNil(tileId)) {
-            if (previousTile?.value === currentTile.value) {
-              score += previousTile.value * 2;
-              newTiles[previousTile.id as string] = {
-                ...previousTile,
-                value: previousTile.value * 2,
-              };
-              newTiles[tileId] = {
-                ...currentTile,
-                position: [newX - 1, y],
-              };
-              previousTile = undefined;
-              hasChanged = true;
-              continue;
-            }
-
-            newBoard[y][newX] = tileId;
-            newTiles[tileId] = {
-              ...currentTile,
-              position: [newX, y],
-            };
-            previousTile = newTiles[tileId];
-            if (!isEqual(currentTile.position, [newX, y])) {
-              hasChanged = true;
-            }
-            newX++;
-          }
-        }
+      if (state.status !== "ongoing") {
+        return state;
       }
-      return {
-        ...state,
-        board: newBoard,
-        tiles: newTiles,
-        hasChanged,
-        score,
-      };
+
+      return moveLeft(state);
     }
     case "move_right": {
-      const newBoard = createBoard();
-      const newTiles: TileMap = {};
-      let hasChanged = false;
-      let { score } = state;
-
-      for (let y = 0; y < tileCountPerDimension; y++) {
-        let newX = tileCountPerDimension - 1;
-        let previousTile: Tile | undefined;
-
-        for (let x = tileCountPerDimension - 1; x >= 0; x--) {
-          const tileId = state.board[y][x];
-          const currentTile = state.tiles[tileId];
-
-          if (!isNil(tileId)) {
-            if (previousTile?.value === currentTile.value) {
-              score += previousTile.value * 2;
-              newTiles[previousTile.id as string] = {
-                ...previousTile,
-                value: previousTile.value * 2,
-              };
-              newTiles[tileId] = {
-                ...currentTile,
-                position: [newX + 1, y],
-              };
-              previousTile = undefined;
-              hasChanged = true;
-              continue;
-            }
-
-            newBoard[y][newX] = tileId;
-            newTiles[tileId] = {
-              ...state.tiles[tileId],
-              position: [newX, y],
-            };
-            previousTile = newTiles[tileId];
-            if (!isEqual(currentTile.position, [newX, y])) {
-              hasChanged = true;
-            }
-            newX--;
-          }
-        }
+      if (state.status !== "ongoing") {
+        return state;
       }
-      return {
-        ...state,
-        board: newBoard,
-        tiles: newTiles,
-        hasChanged,
-        score,
-      };
+
+      return moveRight(state);
     }
     case "reset_game":
       return initialState;
+    case "continue_game":
+      return {
+        ...state,
+        status: "ongoing",
+      };
     case "update_status":
       return {
         ...state,
